hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2423,6 +2425,61 @@ private class Emit(
           }
         }
 
+      case x@NDArraySlice(child, slicesIR) =>
+        val childEmitter = deforest(child)
+
+        val slicest = emit(slicesIR, env, resultRegion, None)
+        val slicesValueAddress = mb.newField[Long]
+        val slicesTuple = new CodePTuple(coerce[PTuple](slicesIR.pType), region, slicesValueAddress)
+
+        val codeSlices = ArrayBuffer[(Code[Long], Code[Long], Code[Long])]()
+        val codeRefMap = mutable.Map[Int, Code[Long]]()
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
+          sliceOrIndex match {
+            case p: PTuple => {
+              val oneSlice = new CodePTuple(p, region, slicesTuple(dim))
+              codeSlices += ((oneSlice[Long](0), oneSlice[Long](1), oneSlice[Long](2)))
+            }
+            case _: PInt64 => {
+              codeRefMap += dim -> slicesTuple(dim)
+            }
+          }
+        }
+
+        val outputShape = codeSlices.toArray.map{ case (start, stop, step) =>
+          (step >= 0L && start <= stop).mux(
+            const(1L) + ((stop - start) - const(1L)) / step,
+            (step < 0L && stop <= start).mux((((stop - start) + const(1L)) / step) + const(1L),
```

Comment:
`[-4:-1:2]`
=>
```
((-1 + 4) + 1)/2 + 1
= 4/2 + 1
= 3
```
seems wrong, right?

Reply:
We should get just -4 and -2.
```
In [20]: [1,2,3,4,5][-4:-1:2]                                                                                                                                                
Out[20]: [2, 4]
```

Reply:
In the example you give, `[1, 2, 3, 4, 5]`, the length is 5, so in python I convert the slice `[-4:-1:2]` to `[1:4:2]`, which will work. That's line 3524 of typed_expressions.py

Reply:
So how do we get `stop <= start` in Scala?

Reply:
Something like `[5:1:-1]`